File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/dashboard/pipeline_representer.rb
##########################################################################
# Copyright 2016 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV1
  module Dashboard
    class PipelineRepresenter < ApiV1::BaseRepresenter
      alias_method :pipeline, :represented

      link :self do |opts|
        opts[:url_builder].pipeline_history_url(pipeline.getName())
      end

      link :doc do
        'https://api.go.cd/#pipelines'
      end

      link :settings_path do |opts|
        opts[:url_builder].pipeline_edit_url(pipeline.getName(), current_tab: :'general')
      end

      link :trigger do |opts|
        opts[:url_builder].api_pipeline_action_url(pipeline.getName(), action: :'schedule')
      end

      link :trigger_with_options do |opts|
        opts[:url_builder].api_pipeline_action_url(pipeline.getName(), action: :'schedule')
      end

      link :pause do |opts|
        opts[:url_builder].pause_pipeline_url(pipeline.getName())
      end

      link :unpause do |opts|
        opts[:url_builder].unpause_pipeline_url(pipeline.getName())
      end

      property :getName, as: :name
      property :locked, exec_context: :decorator
      property :getPausedInfo, as: :pause_info do
        property :paused, as: :paused
        property :pauseBy,
                 as:         :paused_by,
                 getter:     lambda { |options| pauseBy.blank? ? nil : pauseBy },
                 render_nil: true
        property :pauseCause,
                 as:         :pause_reason,
                 getter:     lambda { |options| pauseCause.blank? ? nil : pauseCause },
                 render_nil: true
      end
      collection :instances, embedded: true, exec_context: :decorator, decorator: PipelineInstanceRepresenter

      def locked
        pipeline.getLatestPipelineInstance().isCurrentlyLocked
      end

      def instances
        pipeline.getActivePipelineInstances().select do |pipeline_instance_model|
          !pipeline_instance_model.instance_of?(com.thoughtworks.go.presentation.pipelinehistory.EmptyPipelineInstanceModel)
        end
      end

    end
  end
end

File: api/api-dashboard-v4/src/main/java/com/thoughtworks/go/apiv4/dashboard/representers/PipelineRepresenter.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.apiv4.dashboard.representers;

import com.thoughtworks.go.api.base.OutputLinkWriter;
import com.thoughtworks.go.api.base.OutputListWriter;
import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.config.TrackingTool;
import com.thoughtworks.go.config.remote.RepoConfigOrigin;
import com.thoughtworks.go.domain.PipelinePauseInfo;
import com.thoughtworks.go.presentation.pipelinehistory.EmptyPipelineInstanceModel;
import com.thoughtworks.go.server.dashboard.GoDashboardPipeline;
import com.thoughtworks.go.server.domain.Username;
import com.thoughtworks.go.spark.Routes;
import org.apache.commons.lang3.StringUtils;

import java.util.function.Consumer;

public class PipelineRepresenter {

    public static void toJSON(OutputWriter jsonOutputWriter, GoDashboardPipeline model, Username username) {
        String usernameString = username.getUsername().toString();

        boolean isConfiguredFromConfigRepo = !model.isLocal();
        jsonOutputWriter
                .addLinks(linksWriter -> addLinks(linksWriter, model))
                .add("name", model.name().toString())
                .add("last_updated_timestamp", model.getLastUpdatedTimeStamp())
                .add("locked", model.model().getLatestPipelineInstance().isCurrentlyLocked())
                .addChild("pause_info", getPauseInfoNEW(model))
                .add("can_operate", model.isPipelineOperator(usernameString))
                .add("can_administer", model.canBeAdministeredBy(usernameString))
                .add("can_unlock", model.canBeOperatedBy(usernameString))
                .add("can_pause", model.canBeOperatedBy(usernameString))
                .add("from_config_repo", isConfiguredFromConfigRepo);

        if (isConfiguredFromConfigRepo) {
            RepoConfigOrigin configRepo = (RepoConfigOrigin) model.getOrigin();
            jsonOutputWriter
                    .add("config_repo_id", configRepo.getConfigRepo().getId())
                    .add("config_repo_material_url", configRepo.getMaterial().getUriForDisplay());
        }

        if (model.getTrackingTool().isPresent()) {
            TrackingTool trackingTool = model.getTrackingTool().get();

            jsonOutputWriter.addChild("tracking_tool", childWriter -> {
                childWriter
                        .add("regex", trackingTool.getRegex())
                        .add("link", trackingTool.getLink());
            });
        }

        jsonOutputWriter.addChild("_embedded", childWriter -> {
            childWriter.addChildList("instances", writeInstances(model));
        });
    }

    private static Consumer<OutputListWriter> writeInstances(GoDashboardPipeline model) {
        return listWriter -> {
            model.model().getActivePipelineInstances().stream()
                    .filter(instanceModel -> !(instanceModel instanceof EmptyPipelineInstanceModel))
                    .forEach(instanceModel -> {
                        listWriter.addChild(childWriter -> PipelineInstanceRepresenter.toJSON(childWriter, instanceModel));
                    });
        };
    }

    private static Consumer<OutputWriter> getPauseInfoNEW(GoDashboardPipeline model) {
        return writer -> {
            PipelinePauseInfo pausedInfo = model.model().getPausedInfo();
            writer.add("paused", pausedInfo.isPaused());
            writer.add("paused_by", StringUtils.isBlank(pausedInfo.getPauseBy()) ? null : pausedInfo.getPauseBy());
            writer.add("pause_reason", StringUtils.isBlank(pausedInfo.getPauseCause()) ? null : pausedInfo.getPauseCause());
            writer.add("paused_at", pausedInfo.getPausedAt());
        };
    }

    private static void addLinks(OutputLinkWriter linksWriter, GoDashboardPipeline model) {
        String pipelineName = model.name().toString();
        linksWriter.addLink("self", Routes.PipelineInstance.history(pipelineName))
                .addAbsoluteLink("doc", Routes.Pipeline.DOC);
    }
}
